src/compiler/parser.ts

```diff
@@ -2185,10 +2206,14 @@ namespace ts {
         function parseTypeQuery(): TypeQueryNode {
             const node = <TypeQueryNode>createNode(SyntaxKind.TypeQuery);
             parseExpected(SyntaxKind.TypeOfKeyword);
-            node.exprName = parseEntityName(/*allowReservedWords*/ true);
+            node.exprName = doInExpressionContext(parseEntityNameWithReservedWords);
             return finishNode(node);
         }
 
+        function parseEntityNameWithReservedWords() {

```

Comment:
We normally use named functions in scope for this kind of thing over inline lambdas elsewhere - why change style preference now? I've also been told there's perf differences (in memory usage) between inline function expressions and function declarations; but that could be heresay - feel free to justify one over the other.

Reply:
I wrote a quick tslint rule and there are 2554 arrow functions in our code base, so we're definitely not forbidding them!
```ts
import * as Lint from "tslint/lib";
import * as ts from "typescript";

let count = 0;

export class Rule extends Lint.Rules.AbstractRule {
    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
        ts.forEachChild(sourceFile, function recur(node) {
            if (node.kind === ts.SyntaxKind.ArrowFunction) {
                count++;
            }
            ts.forEachChild(node, recur);
        });
        console.log(count);
        return [];
    }
}
```
Arrow functions make code significantly neater vs named functions since it is immediately obvious that the function is used in only one place, and you don't have to jump around to find out what's done inside a callback.

Reply:
I think it's mostly been a perf-vs-style consideration in each location - we use named functions in perf-sensitive places and tight loops like much of the checker, the transformer, and others, while we use arrows where the code is unlikely to be performance critical (or minimally, places where closures are unlikely to need to be be allocated often). Would the parser be considered performance critical?

Reply:
It's all performance critical, but it's a closure either way, you may be thinking of:
```
function f() {
    foo(() => {});
}
```
vs:
```
function f() {
    foo(g);
}
function g() {}
```

Theoretically, the latter allocates `g` once, the former does it once per call to `f`.
The following offers no advantage over the first way:
```ts
function f() {
    foo(g);
    function g() {}
}
```

Reply:
No, I don't. Given the same containing code, [a function declaration is always faster than an expression](https://jsperf.com/fn-expression-vs-statement), including [in this case](https://jsperf.com/function-expression-capture-vs-declaration-capture). Depending on runtime, the difference can be negligible, but in the past it has been very noticeable.

Reply:
Could you try running this on your computer in node v8.4.0?
```
/*
arrow  fn: 0.00013697601931002756ms
declare fn: 0.00014104301711814795ms
*/

const Benchmark = require("benchmark");
const suite = new Benchmark.Suite;

suite.add("arrow  fn", () => {
    return [1, 2].map(x => x + 1);
});

suite.add("declare fn", () => {
    return [1, 2].map(addOne);
    function addOne(x) {
        return x + 1;
    }
});

suite.on("complete", function() {
    this.forEach(({name, stats}) => {
        console.log(`${name}: ${stats.mean * 1000}ms`);
    });
});
suite.on("error", err => {
    throw err.target.error;
});
suite.run();
```

Reply:
It may randomly give results favorable to "declare fn", but it seems random which one wins.